File: phases/01-math-foundations/03-matrix-transformations/code/transformations.jl
using LinearAlgebra


function rotation_2d(theta)
    c, s = cos(theta), sin(theta)
    return [c -s; s c]
end


function rotation_3d_z(theta)
    c, s = cos(theta), sin(theta)
    return [c -s 0; s c 0; 0 0 1]
end


function rotation_3d_x(theta)
    c, s = cos(theta), sin(theta)
    return [1 0 0; 0 c -s; 0 s c]
end


function rotation_3d_y(theta)
    c, s = cos(theta), sin(theta)
    return [c 0 s; 0 1 0; -s 0 c]
end


function scaling_2d(sx, sy)
    return [sx 0; 0 sy]
end


function shearing_2d(kx, ky)
    return [1 kx; ky 1]
end


function demo_basic_transformations()
    println("=" ^ 60)
    println("基本变换")
    println("=" ^ 60)

    point = [1.0, 0.0]
    theta = pi / 4

    rotated = rotation_2d(theta) * point
    println("\n将 (1,0) 旋转 45 度: $(round.(rotated, digits=4))")

    scaled = scaling_2d(2, 3) * [1.0, 1.0]
    println("将 (1,1) 按 (2,3) 缩放: $(round.(scaled, digits=4))")

    sheared = shearing_2d(1, 0) * [1.0, 1.0]
    println("剪切 (1,1) kx=1: $(round.(sheared, digits=4))")

    reflected = [-1 0; 0 1] * [2.0, 1.0]
    println("沿 y 轴反射 (2,1): $(round.(reflected, digits=4))")
end


function demo_unit_square()
    println("\n" * "=" ^ 60)
    println("单位正方形上的变换")
    println("=" ^ 60)

    square = [[0.0, 0.0], [1.0, 0.0], [1.0, 1.0], [0.0, 1.0]]
    labels = ["origin", "right", "top-right", "top"]

    println("\n原始正方形:")
    for (label, pt) in zip(labels, square)
        println("  $label: $pt")
    end

    transforms = [
        ("Rotate 45 deg", rotation_2d(pi / 4)),
        ("Scale (2, 0.5)", scaling_2d(2, 0.5)),
        ("Shear kx=0.5", shearing_2d(0.5, 0)),
        ("Reflect y-axis", [-1 0; 0 1]),
    ]

    for (name, M) in transforms
        println("\n$name:")
        for (label, pt) in zip(labels, square)
            result = M * pt
            println("  $label: $pt -> $(round.(result, digits=4))")
        end
        println("  det = $(round(det(M), digits=4))")
    end
end


function demo_composition()
    println("\n" * "=" ^ 60)
    println("变换的组合")
    println("=" ^ 60)

    R = rotation_2d(pi / 2)
    S = scaling_2d(2, 0.5)

    point = [1.0, 0.0]

    result1 = (S * R) * point
    result2 = (R * S) * point

    println("\n点: $point")
    println("先旋转 90 再缩放 (2, 0.5): $(round.(result1, digits=4))")
    println("先缩放 (2, 0.5) 再旋转 90: $(round.(result2, digits=4))")
    println("顺序很重要。")

    println("\ndet(R) = $(round(det(R), digits=4))")
    println("det(S) = $(round(det(S), digits=4))")
    println("det(S * R) = $(round(det(S * R), digits=4))")
    println("det(S) * det(R) = $(round(det(S) * det(R), digits=4))")
end


function demo_3d_rotations()
    println("\n" * "=" ^ 60)
    println("3D 旋转")
    println("=" ^ 60)

    point = [1.0, 0.0, 0.0]
    theta = pi / 2

    rz = rotation_3d_z(theta) * point
    rx = rotation_3d_x(theta) * point
    ry = rotation_3d_y(theta) * point

    println("\n点: $point")
    println("绕 z 轴旋转 90: $(round.(rz, digits=4))")
    println("绕 x 轴旋转 90: $(round.(rx, digits=4))")
    println("绕 y 轴旋转 90: $(round.(ry, digits=4))")

    println("\ndet(Rz) = $(round(det(rotation_3d_z(theta)), digits=4))")
    println("det(Rx) = $(round(det(rotation_3d_x(theta)), digits=4))")
    println("det(Ry) = $(round(det(rotation_3d_y(theta)), digits=4))")
    println("所有旋转矩阵的行列式 = 1。")
end


function demo_eigenvalues()
    println("\n" * "=" ^ 60)
    println("特征值和特征向量")
    println("=" ^ 60)

    matrices = [
        ("Symmetric", [2 1; 1 2]),
        ("Upper triangular", [3 1; 0 2]),
        ("Scaling", [3 0; 0 5]),
        ("Rotation 90", [0 -1; 1 0]),
    ]

    for (name, A) in matrices
        vals = eigvals(A)
        vecs = eigvecs(A)
        println("\n$name: $A")
        println("  Eigenvalues: $vals")

        if all(isreal, vals)
            for i in 1:length(vals)
                v = real.(vecs[:, i])
                lam = real(vals[i])
                println("  λ=$(round(lam, digits=4)), v=$(round.(v, digits=4))")
                println("    A * v = $(round.(A * v, digits=4))")
                println("    λ * v = $(round.(lam * v, digits=4))")
            end
        else
            println("  复数特征值：纯旋转，无实特征向量。")
        end
    end
end


function demo_eigendecomposition()
    println("\n" * "=" ^ 60)
    println("特征分解")
    println("=" ^ 60)

    A = Float64[3 1; 0 2]
    F = eigen(A)

    println("\nA = $A")
    println("Eigenvalues: $(F.values)")
    println("Eigenvectors (列):")
    display(F.vectors)
    println()

    V = F.vectors
    D = Diagonal(F.values)
    reconstructed = V * D * inv(V)
    println("重构 A = V * D * V^-1:")
    display(round.(reconstructed, digits=4))
    println()
end


function demo_determinant_meaning()
    println("\n" * "=" ^ 60)
    println("行列式作为体积缩放因子")
    println("=" ^ 60)

    cases = [
        ("Rotation 45 deg", rotation_2d(pi / 4)),
        ("Scale (2, 3)", scaling_2d(2, 3)),
        ("Shear kx=1", shearing_2d(1, 0)),
        ("Reflect y-axis", [-1 0; 0 1]),
        ("Singular", [1 2; 2 4]),
    ]

    println()
    for (name, M) in cases
        d = det(M)
        if abs(d) < 1e-10
            meaning = "空间坍缩，不可逆"
        elseif d < 0
            meaning = "方向翻转"
        elseif abs(d - 1.0) < 1e-10
            meaning = "面积不变"
        else
            meaning = "面积缩放 $(round(abs(d), digits=1)) 倍"
        end
        println("det($name) = $(round(d, digits=4))  ($meaning)")
    end
end


function demo_pca_preview()
    println("\n" * "=" ^ 60)
    println("PCA 预览: 协方差矩阵的特征向量")
    println("=" ^ 60)

    cov = [2.0 1.0; 1.0 3.0]
    F = eigen(cov)

    println("\nCovariance matrix (协方差矩阵): $cov")
    println("Eigenvalues (每个主成分的方差): $(F.values)")
    println("Eigenvectors (主成分):")
    display(F.vectors)
    println()
    println("PCA 选择具有最大特征值的特征向量。")
    println("这里，PC1 捕获了 $(round(F.values[2] / sum(F.values) * 100, digits=1))% 的方差。")
end


demo_basic_transformations()
demo_unit_square()
demo_composition()
demo_3d_rotations()
demo_eigenvalues()
demo_eigendecomposition()
demo_determinant_meaning()
demo_pca_preview()

File: phases/01-math-foundations/03-matrix-transformations/code/transformations.py
import math


def rotation_2d(theta):
    c, s = math.cos(theta), math.sin(theta)
    return [[c, -s], [s, c]]


def rotation_3d_z(theta):
    c, s = math.cos(theta), math.sin(theta)
    return [[c, -s, 0], [s, c, 0], [0, 0, 1]]


def rotation_3d_x(theta):
    c, s = math.cos(theta), math.sin(theta)
    return [[1, 0, 0], [0, c, -s], [0, s, c]]


def rotation_3d_y(theta):
    c, s = math.cos(theta), math.sin(theta)
    return [[c, 0, s], [0, 1, 0], [-s, 0, c]]


def scaling_2d(sx, sy):
    return [[sx, 0], [0, sy]]


def shearing_2d(kx, ky):
    return [[1, kx], [ky, 1]]


def reflection_x():
    return [[1, 0], [0, -1]]


def reflection_y():
    return [[-1, 0], [0, 1]]


def mat_vec_mul(matrix, vector):
    return [
        sum(matrix[i][j] * vector[j] for j in range(len(vector)))
        for i in range(len(matrix))
    ]


def mat_mul(a, b):
    rows_a, cols_b = len(a), len(b[0])
    cols_a = len(a[0])
    return [
        [sum(a[i][k] * b[k][j] for k in range(cols_a)) for j in range(cols_b)]
        for i in range(rows_a)
    ]


def det_2x2(m):
    return m[0][0] * m[1][1] - m[0][1] * m[1][0]


def det_3x3(m):
    return (
        m[0][0] * (m[1][1] * m[2][2] - m[1][2] * m[2][1])
        - m[0][1] * (m[1][0] * m[2][2] - m[1][2] * m[2][0])
        + m[0][2] * (m[1][0] * m[2][1] - m[1][1] * m[2][0])
    )


def eigenvalues_2x2(matrix):
    a, b = matrix[0]
    c, d = matrix[1]
    trace = a + d
    det = a * d - b * c
    discriminant = trace ** 2 - 4 * det
    if discriminant < 0:
        real = trace / 2
        imag = (-discriminant) ** 0.5 / 2
        return (complex(real, imag), complex(real, -imag))
    sqrt_disc = discriminant ** 0.5
    return ((trace + sqrt_disc) / 2, (trace - sqrt_disc) / 2)


def eigenvector_2x2(matrix, eigenvalue):
    a, b = matrix[0]
    c, d = matrix[1]
    if abs(b) > 1e-10:
        v = [b, eigenvalue - a]
    elif abs(c) > 1e-10:
        v = [eigenvalue - d, c]
    else:
        if abs(a - eigenvalue) < 1e-10:
            v = [1, 0]
        else:
            v = [0, 1]
    mag = (v[0] ** 2 + v[1] ** 2) ** 0.5
    return [v[0] / mag, v[1] / mag]


def fmt(v, decimals=4):
    if isinstance(v, list):
        return [round(x, decimals) for x in v]
    return round(v, decimals)


def demo_basic_transformations():
    print("=" * 60)
    print("基本变换")
    print("=" * 60)

    point = [1.0, 0.0]
    theta = math.pi / 4

    rotated = mat_vec_mul(rotation_2d(theta), point)
    print(f"\n将 (1,0) 旋转 45 度: {fmt(rotated)}")

    scaled = mat_vec_mul(scaling_2d(2, 3), [1.0, 1.0])
    print(f"将 (1,1) 按 (2,3) 缩放: {fmt(scaled)}")

    sheared = mat_vec_mul(shearing_2d(1, 0), [1.0, 1.0])
    print(f"剪切 (1,1) kx=1: {fmt(sheared)}")

    reflected = mat_vec_mul(reflection_y(), [2.0, 1.0])
    print(f"沿 y 轴反射 (2,1): {fmt(reflected)}")

    reflected_x = mat_vec_mul(reflection_x(), [2.0, 1.0])
    print(f"沿 x 轴反射 (2,1): {fmt(reflected_x)}")


def demo_unit_square():
    print("\n" + "=" * 60)
    print("单位正方形上的变换")
    print("=" * 60)

    square = [[0, 0], [1, 0], [1, 1], [0, 1]]
    labels = ["origin", "right", "top-right", "top"]

    print("\n原始正方形:")
    for label, pt in zip(labels, square):
        print(f"  {label}: {pt}")

    transforms = [
        ("Rotate 45 deg", rotation_2d(math.pi / 4)),
        ("Scale (2, 0.5)", scaling_2d(2, 0.5)),
        ("Shear kx=0.5", shearing_2d(0.5, 0)),
        ("Reflect y-axis", reflection_y()),
    ]

    for name, matrix in transforms:
        print(f"\n{name}:")
        for label, pt in zip(labels, square):
            result = mat_vec_mul(matrix, pt)
            print(f"  {label}: {pt} -> {fmt(result)}")
        print(f"  det = {fmt(det_2x2(matrix))}")


def demo_composition():
    print("\n" + "=" * 60)
    print("变换的组合")
    print("=" * 60)

    R = rotation_2d(math.pi / 2)
    S = scaling_2d(2, 0.5)

    rotate_then_scale = mat_mul(S, R)
    scale_then_rotate = mat_mul(R, S)

    point = [1.0, 0.0]

    result1 = mat_vec_mul(rotate_then_scale, point)
    result2 = mat_vec_mul(scale_then_rotate, point)

    print(f"\n点: {point}")
    print(f"先旋转 90 再缩放 (2, 0.5): {fmt(result1)}")
    print(f"先缩放 (2, 0.5) 再旋转 90: {fmt(result2)}")
    print("顺序很重要。")

    print(f"\ndet(R) = {fmt(det_2x2(R))}")
    print(f"det(S) = {fmt(det_2x2(S))}")
    print(f"det(S @ R) = {fmt(det_2x2(rotate_then_scale))}")
    print(f"det(S) * det(R) = {fmt(det_2x2(S) * det_2x2(R))}")
    print("组合变换的行列式 = 各行列式的乘积。")


def demo_3d_rotations():
    print("\n" + "=" * 60)
    print("3D 旋转")
    print("=" * 60)

    point = [1.0, 0.0, 0.0]
    theta = math.pi / 2

    rz = mat_vec_mul(rotation_3d_z(theta), point)
    rx = mat_vec_mul(rotation_3d_x(theta), point)
    ry = mat_vec_mul(rotation_3d_y(theta), point)

    print(f"\n点: {point}")
    print(f"绕 z 轴旋转 90: {fmt(rz)}")
    print(f"绕 x 轴旋转 90: {fmt(rx)}")
    print(f"绕 y 轴旋转 90: {fmt(ry)}")

    print(f"\ndet(Rz) = {fmt(det_3x3(rotation_3d_z(theta)))}")
    print(f"det(Rx) = {fmt(det_3x3(rotation_3d_x(theta)))}")
    print(f"det(Ry) = {fmt(det_3x3(rotation_3d_y(theta)))}")
    print("所有旋转矩阵的行列式 = 1（体积不变）。")


def demo_eigenvalues_from_scratch():
    print("\n" + "=" * 60)
    print("特征值和特征向量（从零构建，2x2）")
    print("=" * 60)

    matrices = [
        ("Symmetric", [[2, 1], [1, 2]]),
        ("Upper triangular", [[3, 1], [0, 2]]),
        ("Scaling", [[3, 0], [0, 5]]),
        ("Rotation 90", [[0, -1], [1, 0]]),
    ]

    for name, A in matrices:
        vals = eigenvalues_2x2(A)
        print(f"\n{name}: {A}")
        print(f"  Eigenvalues: {vals[0]}, {vals[1]}")

        if all(isinstance(v, (int, float)) for v in vals):
            for val in vals:
                vec = eigenvector_2x2(A, val)
                result = mat_vec_mul(A, vec)
                scaled = [val * vec[0], val * vec[1]]
                print(f"  λ={fmt(val)}, v={fmt(vec)}")
                print(f"    A @ v = {fmt(result)}")
                print(f"    λ * v = {fmt(scaled)}")
        else:
            print("  复数特征值：纯旋转，无实特征向量。")


def demo_eigendecomposition():
    print("\n" + "=" * 60)
    print("特征分解（2x2，从零构建）")
    print("=" * 60)

    A = [[3, 1], [0, 2]]
    vals = eigenvalues_2x2(A)

    v0 = eigenvector_2x2(A, vals[0])
    v1 = eigenvector_2x2(A, vals[1])

    V = [[v0[0], v1[0]], [v0[1], v1[1]]]
    D = [[vals[0], 0], [0, vals[1]]]

    det_v = det_2x2(V)
    V_inv = [
        [V[1][1] / det_v, -V[0][1] / det_v],
        [-V[1][0] / det_v, V[0][0] / det_v],
    ]

    reconstructed = mat_mul(mat_mul(V, D), V_inv)

    print(f"\nA = {A}")
    print(f"Eigenvalues (特征值): {fmt(vals[0])}, {fmt(vals[1])}")
    print(f"V（特征向量作为列）:")
    for row in V:
        print(f"  {fmt(row)}")
    print(f"D（特征值在对角线上）:")
    for row in D:
        print(f"  {fmt(row)}")
    print(f"重构 A = V @ D @ V^-1:")
    for row in reconstructed:
        print(f"  {fmt(row)}")


def demo_determinant_meaning():
    print("\n" + "=" * 60)
    print("行列式作为体积缩放因子")
    print("=" * 60)

    cases = [
        ("Rotation 45 deg", rotation_2d(math.pi / 4)),
        ("Scale (2, 3)", scaling_2d(2, 3)),
        ("Shear kx=1", shearing_2d(1, 0)),
        ("Reflect y-axis", reflection_y()),
        ("Singular [[1,2],[2,4]]", [[1, 2], [2, 4]]),
    ]

    print()
    for name, m in cases:
        d = det_2x2(m)
        if d == 0:
            meaning = "空间坍缩，不可逆"
        elif d < 0:
            meaning = "方向翻转"
        elif abs(d - 1.0) < 1e-10:
            meaning = "面积不变"
        else:
            meaning = f"面积缩放 {abs(d):.1f} 倍"
        print(f"det({name}) = {fmt(d):>8}  ({meaning})")


def demo_numpy_comparison():
    print("\n" + "=" * 60)
    print("NumPy 对比")
    print("=" * 60)

    try:
        import numpy as np
    except ImportError:
        print("\nNumPy 未安装。跳过。")
        return

    theta = math.pi / 4
    R = np.array([[math.cos(theta), -math.sin(theta)],
                  [math.sin(theta), math.cos(theta)]])

    point = np.array([1.0, 0.0])
    print(f"\n将 (1,0) 旋转 45 度: {R @ point}")

    A = np.array([[2, 1], [1, 2]], dtype=float)
    eigenvalues, eigenvectors = np.linalg.eig(A)
    print(f"\nA = {A.tolist()}")
    print(f"Eigenvalues (numpy): {eigenvalues}")
    print(f"Eigenvectors (numpy, 列):\n{eigenvectors}")

    for i in range(len(eigenvalues)):
        v = eigenvectors[:, i]
        lam = eigenvalues[i]
        print(f"  A @ v{i} = {A @ v}, λ * v{i} = {lam * v}")

    B = np.array([[3, 1], [0, 2]], dtype=float)
    vals, vecs = np.linalg.eig(B)
    D = np.diag(vals)
    V = vecs
    reconstructed = V @ D @ np.linalg.inv(V)
    print(f"\n特征分解 of {B.tolist()}:")
    print(f"  重构: {reconstructed.tolist()}")

    Rz = np.array(rotation_3d_z(math.pi / 2))
    point_3d = np.array([1.0, 0.0, 0.0])
    print(f"\n3D 将 (1,0,0) 绕 z 轴旋转 90 度: {np.round(Rz @ point_3d, 4)}")

    cov = np.array([[2.0, 1.0], [1.0, 3.0]])
    vals, vecs = np.linalg.eig(cov)
    print(f"\nCovariance matrix (协方差矩阵): {cov.tolist()}")
    print(f"Principal components (主成分): 列 of\n{vecs}")
    print(f"Variance along each (每个特征值): {vals}")
    print("PCA 选择具有最大特征值的特征向量。")


if __name__ == "__main__":
    demo_basic_transformations()
    demo_unit_square()
    demo_composition()
    demo_3d_rotations()
    demo_eigenvalues_from_scratch()
    demo_eigendecomposition()
    demo_determinant_meaning()
    demo_numpy_comparison()
